ga4gh/datarepo.py

```diff
@@ -179,42 +262,826 @@ def __init__(
-        data set.
+    class SchemaVersion(object):
+        """
+        The version of the data repository SQL schema
+        """
+        def __init__(self, versionString):
+            splits = versionString.split('.')
+            assert len(splits) == 2
+            self.major = splits[0]
+            self.minor = splits[1]
+
+        def __str__(self):
+            return "{}.{}".format(self.major, self.minor)
+
+    version = SchemaVersion("0.1")
+    systemKeySchemaVersion = "schemaVersion"
+    systemKeyCreationTimeStamp = "creationTimeStamp"
+
+    def __init__(self, fileName):
+        super(SqlDataRepository, self).__init__()
+        self._dbFilename = fileName
+        # We open the repo in either read or write mode. When we want to
+        # update the repo we open it in write mode. For normal online
+        # server use, we open it in read mode.
+        self._openMode = None
+        # Values filled in using the DB. These will all be None until
+        # we have called load()
+        self._schemaVersion = None
+        self._creationTimeStamp = None
+        # Connection to the DB.
+        self._dbConnection = None
+
+    def _checkWriteMode(self):
+        if self._openMode != MODE_WRITE:
+            raise ValueError("Repo must be opened in write mode")
+
+    def _checkReadMode(self):
+        if self._openMode != MODE_READ:
+            raise ValueError("Repo must be opened in read mode")
+
+    def open(self, mode=MODE_READ):
+        """
+        Opens this repo in the specified mode.
+
+        TODO: figure out the correct semantics of this and document
+        the intended future behaviour as well as the current
+        transitional behaviour.
+        """
+        if mode not in [MODE_READ, MODE_WRITE]:
+            error = "Open mode must be '{}' or '{}'".format(
+                MODE_READ, MODE_WRITE)
+            raise ValueError(error)
+        self._openMode = mode
+        if mode == MODE_READ:
+            self.assertExists()
+        self._safeConnect()
+        # Turn on foreign key constraints.
+        cursor = self._dbConnection.cursor()
+        cursor.execute("PRAGMA foreign_keys = ON;")
+        self._dbConnection.commit()
+        if mode == MODE_READ:
+            # This is part of the transitional behaviour where
+            # we load the whole DB into memory to get access to
+            # the data model.
+            self.load()
+
+    def commit(self):
+        """
+        Commits any changes made to the repo. It is an error to call
+        this function if the repo is not opened in write-mode.
+        """
+        self._checkWriteMode()
+        self._dbConnection.commit()
+
+    def close(self):
+        """
+        Closes this repo.
+        """
+        if self._openMode is None:
+            raise ValueError("Repo already closed")
+        self._openMode = None
+        self._dbConnection.close()
+        self._dbConnection = None
+
+    def verify(self):
```

Comment:
It needs to be much more sophisticated, https://github.com/ga4gh/server/issues/1170